internal/inputs/generator_queries.go

```diff
@@ -0,0 +1,272 @@
+		"The number of round-robin serialization groups. Use this to scale up data generation to multiple processes.")
+
+}
+
+// QueryGenerator is a type of Generator for creating queries to test against a
+// database. The output is specific to the type of database (due to each using
+// different querying techniques, e.g. SQL or REST), but is consumed by TSBS
+// query runners like tsbs_run_queries_timescaledb.
+type QueryGenerator struct {
+	// Out is the writer where data should be written. If nil, it will be
+	// os.Stdout unless File is specified in the GeneratorConfig passed to
+	// Generate.
+	Out io.Writer
+	// Out2 is where non-generated data should be written. If nil, it will be
+	// os.Stderr.
+	Out2 io.Writer
+
+	config        *QueryGeneratorConfig
+	useCaseMatrix map[string]map[string]utils.QueryFillerMaker
+	tsStart       time.Time
+	tsEnd         time.Time
+
+	// bufOut represents the buffered writer that should actually be passed to
+	// any operations that write out data.
+	bufOut *bufio.Writer
+}
+
+// NewQueryGenerator returns a QueryGenerator that is set up to work with a given
+// useCaseMatrix, which tells it how to generate the given query type for a use
+// case.
+func NewQueryGenerator(useCaseMatrix map[string]map[string]utils.QueryFillerMaker) *QueryGenerator {
+	return &QueryGenerator{
+		useCaseMatrix: useCaseMatrix,
+	}
+}
+
+func (g *QueryGenerator) Generate(config GeneratorConfig) error {
+	err := g.init(config)
+	if err != nil {
+		return err
+	}
+
+	useGen, err := g.getUseCaseGenerator(g.config)
+	if err != nil {
+		return err
+	}
+
+	filler := g.useCaseMatrix[g.config.Use][g.config.QueryType](useGen)
+
+	return g.runQueryGeneration(useGen, filler, g.config)
+}
+
+func (g *QueryGenerator) init(config GeneratorConfig) error {
+	if config == nil {
+		return fmt.Errorf(ErrNoConfig)
+	}
+	switch config.(type) {
+	case *QueryGeneratorConfig:
+	default:
+		return fmt.Errorf(ErrInvalidDataConfig)
+	}
+	g.config = config.(*QueryGeneratorConfig)
+
+	err := g.config.Validate()
+	if err != nil {
+		return err
+	}
+
+	if _, ok := g.useCaseMatrix[g.config.Use]; !ok {
+		return fmt.Errorf(errBadUseFmt, g.config.Use)
+	}
+
+	if _, ok := g.useCaseMatrix[g.config.Use][g.config.QueryType]; !ok {
+		return fmt.Errorf(errBadQueryTypeFmt, g.config.Use, g.config.QueryType)
+	}
+
+	g.tsStart, err = ParseUTCTime(g.config.TimeStart)
+	if err != nil {
+		return fmt.Errorf(errCannotParseTimeFmt, g.config.TimeStart, err)
+	}
+	g.tsEnd, err = ParseUTCTime(g.config.TimeEnd)
+	if err != nil {
+		return fmt.Errorf(errCannotParseTimeFmt, g.config.TimeEnd, err)
+	}
+
+	if g.Out == nil {
+		g.Out = os.Stdout
+	}
+	g.bufOut, err = getBufferedWriter(g.config.File, g.Out)
+	if err != nil {
+		return err
+	}
+
+	if g.Out2 == nil {
+		g.Out2 = os.Stderr
+	}
+
+	return nil
+}
+
+func (g *QueryGenerator) getUseCaseGenerator(c *QueryGeneratorConfig) (utils.DevopsGenerator, error) {
+	var ret utils.DevopsGenerator
```

Comment:
This is probably a holdover from the older incarnation of this tool, but I always find vars with names like `ret` super hard to follow, might be good to change them to `gen` or `generator` while we're here.